### ISSUE TITLE
hello RTE woes on Mac: memsize, sigfpe, sigsegv

### ISSUE THREAD

**Issue by thorehusfeldt** [RagnarGrootKoerkamp/BAPCtools#281]
Not sure this should be fixed, but neither of `hello/submissions/run_time_error/{test-memsize.cc, test-sigfpe.c, test-sigsegv.c}` behave as expected on my Mac, leading to `pytest` failing already on the first problem.

For `memsize`, I haven’t found a way to limit `ulimit -v` (the size of virtual memory), so the machine blissfully keeps allocating memory for 5 seconds and then `TLE`s. This would be fixable if I understood how to limit virtual memory on my machine.

For `sigfpe`, it turns out that the behaviour of division by 0 even on ints is _undefined_, and some googling shows me that `gcc` on a mac indeed fails to crash also for other people. Maybe this particular submission might be better in a language where `7/0` is _guaranteed_ to crash. (`test-sigfpe.java` crashes just fine.)

For `sigsegv`, my `gcc`  seems to do something clever also at runtime (the compiler already warns about the `-Warray-bounds` violation). No run time error is produced.

**Comment by thorehusfeldt** [comment]
I don’t think `boolfind` can be fixed to ever work on a mac. `clang` fails to link `-lcrt0`, and this is a known frustration with no solution. (As far as I understand.)

`guess` _kinda_ works but results in an impressive array of unexpected failures. I will spend some time looking at these, here’s the full list: (Not super easy to read because the colours are missing.)
```bash
➜  guess git:(ragnar) ✗ bt run
PROBLEM guess
Generate: Done
Running accepted/guess.cc:                                         0.338s ACCEPTED             @ secret/02
Running accepted/guess.java:                                       0.187s ACCEPTED             @ secret/01

Running wrong_answer/GuessCrash.java: secret/02                    0.173s RUN_TIME_ERROR
Exception in thread "main" java.util.NoSuchElementException
	at java.base/java.util.Scanner.throwFor(Scanner.java:937)
	at java.base/java.util.Scanner.next(Scanner.java:1478)
	at GuessCrash.main(GuessCrash.java:10)
judgemessage.txt:
I'm thinking of 1
Guess 1 is 500
Guess 2 is out of range: -1
Running wrong_answer/GuessCrash.java: secret/01                    0.171s RUN_TIME_ERROR
Exception in thread "main" java.util.NoSuchElementException
	at java.base/java.util.Scanner.throwFor(Scanner.java:937)
	at java.base/java.util.Scanner.next(Scanner.java:1478)
	at GuessCrash.main(GuessCrash.java:10)
judgemessage.txt:
I'm thinking of 500
Guess 1 is 500
Trailing output
Running wrong_answer/GuessCrash.java:                              0.173s RUN_TIME_ERROR       @ secret/02

Running wrong_answer/guess.py:                                     0.050s WRONG_ANSWER         @ secret/04

Running wrong_answer/guess_0.cc: secret/03                         0.009s RUN_TIME_ERROR
judgemessage.txt:
I'm thinking of 1000
Guess 1 is 511
Guess 2 is 767
Guess 3 is 895
Guess 4 is 959
Guess 5 is 991
Guess 6 is out of range: 1007
Running wrong_answer/guess_0.cc:                                   0.160s RUN_TIME_ERROR       @ secret/03

Running wrong_answer/guess_extra_output.cc:                        0.155s WRONG_ANSWER         @ secret/02
Running wrong_answer/guess_random.cc:                              0.158s WRONG_ANSWER         @ secret/01

Running wrong_answer/guess_tle.cc: secret/02                       0.252s RUN_TIME_ERROR
judgemessage.txt:
I'm thinking of 1
Guess 1 is out of range: -1
Running wrong_answer/guess_tle.cc: secret/01                       0.252s RUN_TIME_ERROR
judgemessage.txt:
I'm thinking of 500
Guess 1 is out of range: -1
Running wrong_answer/guess_tle.cc:                                 0.252s RUN_TIME_ERROR       @ secret/02

Running time_limit_exceeded/guess_close_stdout_1.cpp:              2.012s TLE (aborted)        @ secret/02
Running time_limit_exceeded/guess_no_flush.cc:                     2.012s TLE (aborted)        @ secret/01

Running time_limit_exceeded/guess_tle_after_correct.cc: secret/03  0.014s RUN_TIME_ERROR
judgemessage.txt:
I'm thinking of 1000
Guess 1 is 500
Guess 2 is 750
Guess 3 is 875
Guess 4 is 938
Guess 5 is 969
Guess 6 is 985
Guess 7 is 993
Guess 8 is 997
Guess 9 is 999
Guess 10 is 1000
Running time_limit_exceeded/guess_tle_after_correct.cc:            0.169s RUN_TIME_ERROR       @ secret/03

Running run_time_error/guess_assert_failure.cpp:                   0.157s RUN_TIME_ERROR       @ secret/02
Running run_time_error/guess_assert_failure.py:                    0.129s RUN_TIME_ERROR       @ secret/02
Running run_time_error/guess_bad_exit.py:                          0.052s RUN_TIME_ERROR       @ secret/02
Running run_time_error/guess_close_stdout.cpp:                     0.155s RUN_TIME_ERROR       @ secret/02

Running run_time_error/guess_close_stdout_1.cpp: secret/02         1.162s TIME_LIMIT_EXCEEDED
Assertion failed: (false), function main, file guess_close_stdout_1.cpp, line 9.
judgemessage.txt:
I'm thinking of 1
Guess 1: couldn't read an integer
Running run_time_error/guess_close_stdout_1.cpp: secret/01         1.170s TIME_LIMIT_EXCEEDED
Assertion failed: (false), function main, file guess_close_stdout_1.cpp, line 9.
judgemessage.txt:
I'm thinking of 500
Guess 1: couldn't read an integer
Running run_time_error/guess_close_stdout_1.cpp:                   1.170s TIME_LIMIT_EXCEEDED  @ secret/01

Running run_time_error/guess_close_stdout_2.cpp: secret/02         1.166s TIME_LIMIT_EXCEEDED
judgemessage.txt:
I'm thinking of 1
Guess 1: couldn't read an integer
Running run_time_error/guess_close_stdout_2.cpp: secret/01         1.165s TIME_LIMIT_EXCEEDED
judgemessage.txt:
I'm thinking of 500
Guess 1: couldn't read an integer
Running run_time_error/guess_close_stdout_2.cpp:                   1.166s TIME_LIMIT_EXCEEDED  @ secret/02

Running run_time_error/guess_exit_1.cpp:                           0.165s RUN_TIME_ERROR       @ secret/02
Running run_time_error/guess_rte.c:                                0.152s RUN_TIME_ERROR       @ secret/01
Running run_time_error/guess_rte_after_correct.cc:                 0.181s RUN_TIME_ERROR       @ secret/01
```

**Comment by thorehusfeldt** [comment]
(I would like at least a few interesting problems to actually work even on a mac, and to the extent it’s fixable we should do it. I’ll try to come up with a sensible compromise.)

**Comment by RagnarGrootKoerkamp** [comment]
All these problems are when running manually right, not just under `pytest`?

> clang fails to link -lcrt0

What is `crt0`? Grepping the codebase doesn't return anything?

Yes, I agree that basic interactive problems should work. I don't see a reason while they shouldn't work, but clearly I'm missing things. Note that compilating issues in the specific output validators/interactors used for these problems should be fixable.

**Comment by thorehusfeldt** [comment]
Can somebody else verify the behaviour of `guess` for me? For instance, it seems as if `GuessCrash.java` is _correctly_ judged `RTE` (since the interactive verifier sent no output back), so `myString = sc.next();` raises an exception).

Does this boil down to a race between the interactive verifier equally correctly aborting with `Guess 2 is out of range: -1`? How is such a situation defined? (judge failed with `WA`, team crashed with `RTE`?) 

I’m honestly stumped.

**Comment by RagnarGrootKoerkamp** [comment]
Ok so for `gues` the `AC` submissions do work. But many non-AC submissions give wrong verdicts. This is not uncommon in general for interactive problems where deciding the right verdict is actually quite tricky.

Oh in fact, for windows and mac we have much simplified handling of interactive test cases, since we cannot accurately determine which of the submission and interactor exists first without `os.wait3` or `os.wait4`, both of which only exist on linux.
https://github.com/RagnarGrootKoerkamp/BAPCtools/blob/master/bin/interactive.py#L77
https://github.com/RagnarGrootKoerkamp/BAPCtools/blob/master/bin/interactive.py#L264

So basically: on windows/mac, all we can do is decide AC vs non-AC, and all the tricky ways in which programs can get non-AC are too hard to distinguish correctly.

**Comment by RagnarGrootKoerkamp** [comment]
> Does this boil down to a race between the interactive verifier equally correctly aborting with `Guess 2 is out of range: -1`? How is such a situation defined? (judge failed with `WA`, team crashed with `RTE`?)

Yes so another thing that should be done is to write a proper spec interactive problem validation. I believe there is a draft / some notes somewhere, but this is more a 'commonly accepted list' that the kattis/domjudge/bapctools people are aware of than super pinned down. The important part for now is:

- As soon as the interactor decides it's WA, it may exit and return so.
- This can crash the submission, since now their `stdin` is closed. Hence, we just kill it directly after the interactor exits

**Comment by thorehusfeldt** [comment]
(Yes, the problems fail manually using `bt run`, not only from pytest.)

**Comment by thorehusfeldt** [comment]
I cannot run `problemtools` on a mac at all; I’ve long given up compiling it, and know nobody who has succeeded in recent memory. 

(That’s one of the reasons I find BAPCtools attractive: I can hack it) 

So I run `problemtools` exclusively via `docker`.

**Comment by RagnarGrootKoerkamp** [comment]
Anyway, I think a reasonable solution here is to be more more lenient with the expected results in this specific case:
- On mac
- Interactive problem
- Any kind of non-AC is treated 'the same' and everything wrong submission (`submissions/{wa,rte,tle}`) is shown in green as long as they return non-AC.
